backend/src/replication/StatisticsReplication.ts

```diff
@@ -0,0 +1,231 @@
+  inProgress: boolean = false;
+
+  public async $sync(): Promise<void> {
+    if (!config.REPLICATION.ENABLED || !config.REPLICATION.STATISTICS || !config.STATISTICS.ENABLED) {
+      // replication not enabled, or statistics not enabled
+      return;
+    }
+    if (this.inProgress) {
+      logger.info(`StatisticsReplication sync already in progress`, 'Replication');
+      return;
+    }
+    this.inProgress = true;
+
+    const missingStatistics = await this.$getMissingStatistics();
+    const missingIntervals = Object.keys(missingStatistics).filter(key => missingStatistics[key].size > 0);
+    const totalMissing =  missingIntervals.reduce((total, key) => total + missingStatistics[key].size, 0);
+
+    if (totalMissing === 0) {
+      this.inProgress = false;
+      logger.info(`Statistics table is complete, no replication needed`, 'Replication');
+      return;
+    }
+    
+    for (const interval of missingIntervals) {
+      logger.debug(`Missing ${missingStatistics[interval].size} statistics rows in '${interval}' timespan`, 'Replication');
+    }
+    logger.debug(`Fetching ${missingIntervals.join(', ')} statistics endpoints from trusted servers to fill ${totalMissing} rows missing in statistics`, 'Replication');
+    
+    let totalSynced = 0;
+    let totalMissed = 0;
+
+    for (const interval of missingIntervals) {
+      const results = await this.$syncStatistics(interval, missingStatistics[interval]);
+      totalSynced += results.synced;
+      totalMissed += results.missed;
+
+      logger.info(`Found ${totalSynced} / ${totalSynced + totalMissed} of ${totalMissing} missing statistics rows`, 'Replication');
+      await Common.sleep$(3000);
+    }
+
+    logger.debug(`Synced ${totalSynced} statistics rows, ${totalMissed} still missing`, 'Replication');
+
+    this.inProgress = false;
+  }
+
+  private async $syncStatistics(interval: string, missingTimes: Set<number>): Promise<any> {
+  
+    let success = false;
+    let synced = 0;
+    let missed = 0;
+    const syncResult = await $sync(`/api/v1/statistics/${interval}`);
+    if (syncResult) {
+      if (syncResult.data?.length) {
+        success = true;
+        logger.info(`Fetched /api/v1/statistics/${interval} from ${syncResult.server}`);
+        const fetchedTimes = syncResult.data.map((stat: any) => this.roundToNearestStep(stat.added, steps[interval]));
+          for (const time of missingTimes) {
+          const index = fetchedTimes.indexOf(time);
+          if (index !== -1) {
+            const stat = syncResult.data[index];
+            try {
+              await statistics.$create(statistics.mapOptimizedStatisticToStatistic([stat])[0], true);
+              synced++;
+            } catch (e: any) {
+              logger.err(`Failed to insert statistics row at ${time} (${interval}) from ${syncResult.server}. Reason: ` + (e instanceof Error ? e.message : e));
+            }
+          } else {
+            missed++;
+          }
+        }
```

Comment:
this approach works fine, but it does mean we lose a bit of precision in the 2h/24h charts by discarding some of the extra points before and after new blocks (see https://github.com/mempool/mempool/pull/4676).

| <img width="1166" alt="Screenshot 2024-05-21 at 11 26 13 PM" src="https://github.com/mempool/mempool/assets/83316221/7207b34b-759e-4f84-b825-39010cd8de5f"> | vs | <img width="1166" alt="Screenshot 2024-05-21 at 11 28 10 PM" src="https://github.com/mempool/mempool/assets/83316221/a1a00f9a-63b3-4585-bbf8-c5b368760608"> |
|-|-|-|

an alternative would be to iterate over the fetchedTimes instead of the missingTimes, and accept anything within half a step of a missingTime entry. that should still avoid duplicate data, while filling any gaps with the best level of detail.

Reply:
> iterate over the fetchedTimes instead of the missingTimes, and accept anything within half a step of a missingTime entry

Right, doing so fixes the issue: 

Trusted server: 
<img width="1000" alt="Screenshot 2024-05-22 at 18 18 36" src="https://github.com/mempool/mempool/assets/46578910/38c7c63b-d1af-4d5b-aa1a-38d786ed5de8">

Replication (the drop at the end is my local mempool not being the same as the trusted server): 
<img width="1000" alt="Screenshot 2024-05-22 at 18 18 58" src="https://github.com/mempool/mempool/assets/46578910/2e114bda-b79e-48cd-ae50-189a04a155e1">